### PR TITLE
ci(.github): refactor build workflow to integrate sca, signing and provenance for images and artifacts

### PR DESCRIPTION
**Pull request by saisatishkarra**
### Checklist prior to review

- [ ] [Link to relevant issue][1] as well as docs and UI issues --
- [ ] This will not break child repos: it doesn't hardcode values (.e.g "kumahq" as a image registry) and it will work on Windows, system specific functions like `syscall.Mkfifo` have equivalent implementation on the other OS --
- [ ] Tests (Unit test, E2E tests, manual test on universal and k8s) --
  - Don't forget `ci/` labels to run additional/fewer tests
- [ ] Do you need to update [`UPGRADE.md`](../blob/master/UPGRADE.md)? --
- [ ] Does it need to be backported according to the [backporting policy](../blob/master/CONTRIBUTING.md#backporting)? ([this](https://github.com/kumahq/kuma/actions/workflows/auto-backport.yaml) GH action will add "backport" label based on these [file globs](https://github.com/kumahq/kuma/blob/master/.github/workflows/auto-backport.yaml#L6), if you want to prevent it from adding the "backport" label use [no-backport-autolabel](https://github.com/kumahq/kuma/blob/master/.github/workflows/auto-backport.yaml#L8) label) --

### Overview
- Created Signature repositories to push image signatures for verification.
   - `kumahq/notary` public-facing docker repository for release tags
   - `kumahq/notary-internal` for internal docker for internal / preview tags
- Refactors build pipelines to integrate
   -  Reusable docker images TAR/artifact binary files across workflow jobs using upload/download v4 action
   - Scanning (SCA) of docker images
   - Signing of docker image manifests recursively for all platforms and publishing signatures to the docker hub (registry needs to be created)
   - Build provenance for artifacts and images being produced using slsa-generators (generic, container)
      - Provenance is NOT generated on PR events and hence cannot test it until merged 

### TODO
- Test the workflow first on Kumahq and sync changes to Kong-mesh'
- Update any kong-mesh / child repo workflow sync files for the refactored pipeline

### Deps
- PR from GUI with SBOM assets in `data` dir

[1]: https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword

### CODE REVIEW

**Comment by saisatishkarra** [comment]
cc: @kleinfreund Is there a GUI PR in the master to be rebased with SBOM assets to test if  GUI deps are being tracked as part of scanning the Kuma repo?

File: .github/workflows/build.yml [hunk 1]
```diff
@@ -0,0 +1,129 @@
...
+        run: |
+          echo 'ALLOW_PUSH=true' >> $GITHUB_OUTPUT
+      - uses: actions/cache@13aacd865c20de90d75de3b17ebe84f7a17d57d2 # v4.0.0
+        id: ci_tools_cache
+        with:
+          path: |
+            ${{ env.CI_TOOLS_DIR }}
+          key: ${{ runner.os }}-${{ runner.arch }}-devtools-${{ hashFiles('mk/dependencies/deps.lock') }}
+          restore-keys: |
+            ${{ runner.os }}-${{ runner.arch }}-devtools
```
Comment (lahabana) [inline .github/workflows/build.yml:50]: I'm not sure what the last step is for.

**Review by lahabana — COMMENTED** [review]
This is good thanks for splitting things up it's much easier to review

File: .github/workflows/build.yml (hunk 1 above)
Comment (saisatishkarra) [reply .github/workflows/build.yml]: removed this step from `metadata` job

File: .github/workflows/build.yml [hunk 2]
```diff
@@ -0,0 +1,129 @@
...
+        if: github.event_name == 'push' || contains(github.event.pull_request.labels.*.name, 'ci/run-full-matrix')
+        id: set-full-matrix-switches
+        run: |
+          echo 'ENABLED_GOARCHES=arm64 amd64' >> $GITHUB_OUTPUT
+          echo 'ENABLED_GOOSES=linux darwin' >> $GITHUB_OUTPUT
+      - name: "Maybe set flag to push build artifacts"
+        id: artifacts_metadata
+        if: github.event_name == 'push' || contains(github.event.pull_request.labels.*.name, 'ci/force-publish')
+        run: |
+          echo 'ALLOW_PUSH=true' >> $GITHUB_OUTPUT
```
[the remark this answers is not in the document]
Comment (saisatishkarra) [reply .github/workflows/build.yml]: the `metadata` job is moved to `build-test-distribute` and the 3 things are passed as inputs and set as env in `build/provenance/release` jobs which depend on `metadata` job

File: .github/workflows/distributions.yml [hunk 3]
```diff
@@ -0,0 +1,208 @@
...
+      IMAGE_MANIFESTS:
+        value: ${{ jobs.publish-images.outputs.IMAGE_MANIFESTS }}
+env:
+  CI_TOOLS_DIR: /home/runner/work/kuma/kuma/.ci_tools
+  ENABLED_GOARCHES: ${{ inputs.enabled_goarches }}
+  ENABLED_GOOSES: ${{ inputs.enabled_gooses }}
+  ALLOW_PUSH: ${{ inputs.allow_push }}
+  NOTARY_REPOSITORY: "${{ github.repository_owner }}/notary"
+  GH_USER: "github-actions[bot]"
+  GH_EMAIL: "<41898282+github-actions[bot]@users.noreply.github.com>"
```
[the remark this answers is not in the document]
Comment (saisatishkarra) [reply .github/workflows/distributions.yml:45]: done

File: .go-version [hunk 4]
```diff
@@ -0,0 +1 @@
+1.21.5
```
[the remark this answers is not in the document]
Comment (saisatishkarra) [reply .go-version]: removed this file. Was unncessary

File: .github/workflows/build-test-distribute.yaml [hunk 5]
```diff
@@ -49,6 +49,54 @@ jobs:
...
+    needs: ["build"]
+    uses: ./.github/workflows/sca.yml
+    with:
+      images: "${{ needs.build.outputs.IMAGES }}"
+      artifact_name: ${{ needs.build.outputs.IMAGE_ARTIFACTS_NAME }}
+  release:
+    permissions:
+      contents: read
+      id-token: write # needed for signing the images
+    needs: ["build", "scan"]
```
[the remark this answers is not in the document]
Comment (saisatishkarra) [reply .github/workflows/build-test-distribute.yaml]: modified to `needs: ["metadata", "build", "scan", "test", "test_e2e", "test_e2e_env"]`

**Comment by saisatishkarra** [comment]
@lahabana is there an ETA on when the mesh team can review/ restart work on this PR?

**Comment by lahabana** [comment]
@saisatishkarra the PR is being integrated in smaller bit. The massive refactor was hard to integrate as one block. #9360 was merged first now we're doing: https://github.com/kumahq/kuma/pull/9393/ and others will follow

**Comment by lahabana** [comment]
This was done is separate PRs. Closing this now

**Pull request closed without merging**